### PR TITLE
[SDRAN-910] Introducing encoding of REAL structs

### PR DESCRIPTION
**Pull request by eroshiva**
Decoding part is still missing

### CODE REVIEW

**Comment by eroshiva** [comment]
Some more polishing is needed (in decoding part)

File: pkg/asn1/aper/aper.go [hunk 1]
```diff
@@ -419,6 +419,124 @@ func (pd *perBitData) parseBool() (value bool, err error) {
...
+
+	var base int
+	// parsing encoding base
+	bb, err := pd.getBitsValue(2)
+	if err != nil {
+		return 0, err
+	}
+	switch bb {
+	case 0:
+		base = 2
```
Comment (SeanCondon) [inline pkg/asn1/aper/aper.go:469]: base 2 is the only one supported!

File: pkg/asn1/aper/aper.go [hunk 2]
```diff
@@ -419,6 +419,124 @@ func (pd *perBitData) parseBool() (value bool, err error) {
...
+	ff, err := pd.getBitsValue(2)
+	if err != nil {
+		return 0, err
+	}
+	if ff != 0 {
+		return 0, errors.NewInvalid("Error parsing scaling factor - decoded bits expected to be 0, obtained %v", ff)
+	}
+
+	// parsing exponent
+	ee, err := pd.getBitsValue(2)
```
Comment (SeanCondon) [inline pkg/asn1/aper/aper.go:491]: is this exponent length or type or something?

File: pkg/asn1/aper/aper.go [hunk 3]
```diff
@@ -419,6 +419,124 @@ func (pd *perBitData) parseBool() (value bool, err error) {
...
+	mantissa, err := pd.getBitsValue(8 * uint(byteLength-1-1))
+	if err != nil {
+		return 0, err
+	}
+	log.Debugf("Obtained mantissa is %v", mantissa)
+
+	if twosComplement {
+		result = float64(mantissa)
+		for i := 0; i < int(exponent); i++ {
+			result = result / float64(base)
```
Comment (SeanCondon) [inline pkg/asn1/aper/aper.go:522]: nice

File: pkg/asn1/aper/marshal.go [hunk 4]
```diff
@@ -346,6 +346,187 @@ func (pd *perRawBitData) appendBool(value bool) (err error) {
...
+	byteAmount++
+
+	return
+}
+
+// it looks like encoding of REAL doesn't take into account constraints at all, so we don't bother about parsing constraints
+// general rules are - mantissa should be an odd number or a 0
+func (pd *perRawBitData) appendReal(value float64) (err error) {
+
+	log.Debugf("Encoding REAL number %v", value)
```
Comment (SeanCondon) [inline pkg/asn1/aper/marshal.go:391]: amazing

File: pkg/asn1/aper/marshal.go [hunk 5]
```diff
@@ -346,6 +346,187 @@ func (pd *perRawBitData) appendBool(value bool) (err error) {
...
+		// mantissa can't be negative
+		if mantissa < 0 {
+			mantissa = -mantissa
+		}
+		negativeExponent = true
+		log.Debugf("Obtained mantissa is %v, value in computations is %v. Exponent is %v, it is negative (%v)", mantissa, val, exponent, negativeExponent)
+		log.Debugf("Computing 2's complement for exponent (%v)", exponent)
+		//alternative way - works for 8 bit representation numbers
+		twosComplimentExp := 256 - exponent
+		log.Debugf("2's complement for exponent %v is %v", exponent, twosComplimentExp)
```
Comment (SeanCondon) [inline pkg/asn1/aper/marshal.go:444]: you could combine these in to 1 log statement

File: pkg/asn1/aper/marshal.go [hunk 6]
```diff
@@ -346,6 +346,187 @@ func (pd *perRawBitData) appendBool(value bool) (err error) {
...
+		exponent = twosComplimentExp
+		log.Debugf("It requires %v bits to store the value", howManyBitsNeeded(exponent))
+	}
+
+	p = howManyBytesNeeded(mantissa)
+	log.Debugf("Exponent/2 is %v, mantissa/2 is %v, exponent is %v, p is %v, mantissa is %v", exponent%2, mantissa%2, exponent, p, mantissa)
```
Comment (SeanCondon) [inline pkg/asn1/aper/marshal.go:450]: you could say `Exponent%2` or `is odd` or `is even`
Not clear that `p` is the number of bytes

File: pkg/asn1/aper/marshal.go [hunk 7]
```diff
@@ -346,6 +346,187 @@ func (pd *perRawBitData) appendBool(value bool) (err error) {
...
+	if (exponent%2 == 0 || mantissa%2 != 0 || exponent == 0) && p < 7 && mantissa > 32 {
+		// 7 is the maximum number of bytes to carry mantissa (because of max. 51 multiplication of 2),
+		// mantissa > 32 and odd - reverse engineered from Nokia's asn1c tool
+		p = p + 1
+	}
+	n = howManyBytesNeeded(exponent) // should be always 1 byte
+
+	// computing length of the bytes needed to encode a number
+	byteLength := n + p + 1
+	log.Debugf("Amount of bytes to encode is %v: 1 byte for length, 1 byte for header, %v bytes for exponent, %v bytes for mantissa", byteLength, n, p)
```
Comment (SeanCondon) [inline pkg/asn1/aper/marshal.go:463]: when I ran test I get 
```
Amount of bytes to encode is 9: 1 byte for length, 1 byte for header, 1 bytes for exponent, 7 bytes for mantissa
```
does not add up

**Review by SeanCondon — APPROVED** [review]

**Pull request merged**